File: Maximum Number of Points From Grid Queries/Maximum Number of Points From Grid Queries.cpp
// Solution for Maximum Number of Points From Grid Queries in CPP

class Solution {
public:
    vector<int> maxPoints(vector<vector<int>>& grid, vector<int>& queries) {
        int m = grid.size(), n = grid[0].size();
        int q = queries.size();
        vector<int> res(q, 0);
        vector<pair<int, int>> sortedQ;
        for (int i = 0; i < q; i++) {
            sortedQ.push_back({queries[i], i});
        }
        sort(begin(sortedQ), end(sortedQ));
        vector<vector<int>> directions = {{0, 1}, {0, -1}, {1, 0}, {-1, 0}};
        priority_queue<vector<int>, vector<vector<int>>, greater<>> minH;
        vector<vector<int>> vis(m, vector<int>(n, false));
        int points = 0;
        minH.push({grid[0][0], 0, 0});
        vis[0][0] = true;
        for (int i = 0; i < q; i++) {
            int val = sortedQ[i].first;
            int idx = sortedQ[i].second;
            while (!minH.empty() && minH.top()[0] < val) {
                int i = minH.top()[1];
                int j = minH.top()[2];
                minH.pop();
                points++;
                for (const auto& dir: directions) {
                    int i_ = i + dir[0];
                    int j_ = j + dir[1];
                    if (i_ >= 0 && i_ < m && j_ >= 0 && j_ < n && !vis[i_][j_]) {
                        minH.push({grid[i_][j_], i_, j_});
                        vis[i_][j_] = true;
                    }
                }
            }
            res[idx] = points;
        }
        return res;
    }
};

File: Maximum Number of Points From Grid Queries/Maximum Number of Points From Grid Queries.py
# Solution for Maximum Number of Points From Grid Queries in PY

class Solution:
    def maxPoints(self, grid: List[List[int]], queries: List[int]) -> List[int]:
        m, n, q = len(grid), len(grid[0]), len(queries)
        res = [0] * q
        # sortedQ = []
        # for i in range(q):
        #     sortedQ.append([queries[i], i])
        # sortedQ.sort(key = lambda x: x[0])
        sortedQ = sorted([(queries[i], i) for i in range(q)], key = lambda x: x[0])

        directions = [[0, 1], [0, -1], [1, 0], [-1, 0]]
        minH = []
        vis = [[False] * n for _ in range(m)]
        points = 0

        heapq.heappush(minH, (grid[0][0], 0, 0))
        vis[0][0] = True

        for val, idx in sortedQ:
            while minH and minH[0][0] < val:
                _, i, j = heapq.heappop(minH)
                points += 1
                for dir in directions:
                    i_, j_ = i + dir[0], j + dir[1]
                    if 0 <= i_ < m and 0 <= j_ < n and not vis[i_][j_]:
                        heapq.heappush(minH, (grid[i_][j_], i_, j_))
                        vis[i_][j_] = True
            res[idx] = points

        return res
